src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+module Generalize (generalizeArgs, generalizeIxDict, Generalized) where
+
+import Control.Monad
+import Control.Category ((>>>))
+
+import Syntax
+import Inference
+import QueryType
+import Name
+import Err
+import MTL1
+import LabeledItems
+
+type Generalized (e::E) (n::S) = (Abs (Nest Binder) e n, [Atom n])
+
+generalizeIxDict :: EnvReader m => Dict n -> m n (Generalized Dict n)
+generalizeIxDict dict = liftGeneralizerM do
+  dict' <- sinkM dict
+  dictTy <- getType dict'
+  dictTyGeneralized <- generalizeType dictTy
+  liftEnvReaderM $ generalizeDict dictTyGeneralized dict'
+{-# INLINE generalizeIxDict #-}
+
+generalizeArgs ::EnvReader m => Nest PiBinder n l -> [Atom n] -> m n (Generalized (ListE Atom) n)
+generalizeArgs reqTys allArgs =  liftGeneralizerM $ runSubstReaderT idSubst do
+  PairE (Abs reqTys' UnitE) (ListE allArgs') <- sinkM $ PairE (Abs reqTys UnitE) (ListE allArgs)
+  ListE <$> go reqTys' allArgs'
+  where
+    go ::Nest PiBinder i i' -> [Atom n] -> SubstReaderT AtomSubstVal GeneralizerM i n [Atom n]
+    go Empty [] = return []
+    go (Nest (PiBinder b ty _) bs) (arg:args) = do
+      ty' <- substM ty
+      arg' <- case ty' of
+        TyKind   -> liftSubstReaderT $ generalizeType arg
+        DictTy _ -> generalizeDict ty' arg
+        _ -> return arg
```

Comment:
For example, here you do resynthesize the roles based on the types, instead of using the annotated binders. So why do we even need the binder annotations?

Reply:
Right, so you basically do role inference for functions here? What about the point of a dict that's an explicit argument? Perhaps you should take the arrow into account?

Reply:
Something strange seems to be happening with github. Right now, I see this line as:
```
DictTy _ -> generalizeDict ty' arg
```
But it should actually be this:
```
DictTy _ | arr == ClassArrow -> generalizeDict ty' arg
```
I changed it in efd06e9f5ef15a8beb3cce9233ff98d66564d8cd, which is part of this PR and is now merged into main. But for some reason it's not showing up.

The reason I don't want to use the helper `inferRole` from `Inference` is that we need a less strict rule here. It's ok for arguments to functions to be non-data things. We just specialize on the actual value in that case.